### PR TITLE
Add a test for canceling resources with a timeout

### PR DESCRIPTION
**Pull request by iwahbe**
When I was answering https://github.com/pulumi/pulumi-go-provider/issues/261#issuecomment-2688957783, I was looking at our implementation of `Cancel`. I noticed there wasn't a test for this edge case, and added one.

### CODE REVIEW

File: tests/cancel_test.go [hunk 1]
```diff
@@ -181,3 +181,46 @@ func TestCancelTimeout(t *testing.T) {
 		assert.ErrorIs(t, err, context.DeadlineExceeded)
 	})
 }
+
+// Test that a `Cancel` call will cancel in-flight operations, even if they already have a
+// timeout associated with them.
+//
+//	Main routine                   Server routine                 Cancel routine
```
Comment (thomas11) [inline tests/cancel_test.go:188]: A+ comment

**Review by thomas11 — APPROVED** [review]
